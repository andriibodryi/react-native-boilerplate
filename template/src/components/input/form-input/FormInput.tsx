import React from 'react';
import { useController } from 'react-hook-form';

import { FormInputTypes } from './form-input.types';
import { useFormInputOnOnceDirty } from './utils/use-on-once-dirty.hook';
import { useFormInputOnError } from './utils/use-on-error.hook';
import { useCheckRulesOnMount } from './utils/use-check-rules-on-mount.hook';
import { formInputMap } from './form-input.constants';

const FormInput: React.FC<FormInputTypes> = ({
  type,
  name,
  rules,
  defaultValue,
  onOnceDirty,
  onErrorMap: onError,
  checkRulesOnMount,
  ...props
}) => {
  const { field, fieldState } = useController({
    name,
    rules,
    defaultValue,
  });

  useCheckRulesOnMount(field, checkRulesOnMount);
  useFormInputOnOnceDirty(fieldState.isDirty, onOnceDirty);
  useFormInputOnError(fieldState.error, onError);

  const FormInputComponent = formInputMap[type];

  return (
    <FormInputComponent
      onChangeText={field.onChange}
      onBlur={field.onBlur}
      value={field.value}
      errorMessage={fieldState.error?.message}
      {...props}
    />
  );
};

export default FormInput;
