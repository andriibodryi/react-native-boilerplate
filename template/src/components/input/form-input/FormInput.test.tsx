import React from 'react';
import { render } from '@testing-library/react-native';

import FormInput from './FormInput';

const dummyForm = {
  formState: {
    errors: {},
  },
};

jest.mock('react-hook-form', () => ({
  useController: () => ({
    field: {
      value: 'dummyValue',
      onChange: () => {},
      onBlur: () => {},
    },
    fieldState: {
      isDirty: false,
      error: false,
    },
  }),
  useFormContext: () => dummyForm,
}));

describe('FormInput', () => {
  it('should render correctly', () => {
    const { toJSON } = render(<FormInput type="text" name="dummyName" />);

    expect(toJSON()).toMatchSnapshot();
  });
});
