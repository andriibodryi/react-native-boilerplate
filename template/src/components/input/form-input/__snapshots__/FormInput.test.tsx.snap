// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput should render correctly 1`] = `
Array [
  <RCTInputAccessoryView
    nativeID="testNativeID"
    style={
      Array [
        undefined,
        Object {
          "position": "absolute",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "padding": 4,
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0.5,
          },
          Object {
            "backgroundColor": "#F2F2F2",
          },
          undefined,
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "rgb(127, 43, 123)",
                },
              ]
            }
          >
            Done
          </Text>
        </View>
      </View>
    </View>
  </RCTInputAccessoryView>,
  <View
    style={
      Object {
        "marginBottom": 4,
      }
    }
  />,
  <View>
    <View
      style={
        Array [
          Object {
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "space-between",
            "overflow": "hidden",
            "paddingLeft": 15,
            "position": "relative",
          },
          Object {
            "borderColor": "#C3C3C3",
          },
          false,
          false,
          Object {
            "alignItems": "center",
          },
          undefined,
        ]
      }
    >
      <TextInput
        accessibilityHint="Double tap to enter"
        accessible={true}
        allowFontScaling={true}
        editable={true}
        inputAccessoryViewID="testNativeID"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="#808080"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "flex": 1,
              "fontFamily": "Aspira-Regular",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 20,
              "position": "relative",
              "top": -2,
            },
            Object {
              "color": "#4D4D4D",
            },
            undefined,
          ]
        }
        testID="input_field_dummyname"
        underlineColorAndroid="transparent"
        value="dummyValue"
      />
    </View>
    <View
      style={
        Object {
          "marginTop": 6,
          "minHeight": 16,
          "position": "relative",
        }
      }
    />
  </View>,
]
`;
